### PR TITLE
[4.0] [CSP] Initial Integration of com_csp & plg_system_httpheaders

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

This PR is the next step for the implementation of a core csp handling.
This now does the integration of com_csp & plg_system_httpheaders.

The general goal of the PR is that everything that is published in com_csp also gets added to the automatic csp rule including some general things like nonce's

##### Testing the `detecting` mode
- Go to the Plugin
- Got to the CSP tab in the plugin settings
- Set the `mode` to `detecting`
- Visit the frontend & navigate in the backend
- check the com_csp backend
- notice the collected reports also with client info

##### Testing the `custom` mode

- Go to the Plugin
- Got to the CSP tab in the plugin settings
- Set the `mode` to `custom`
- Notice that now the additional fields open
- configure (i would suggest) a report only rule
- confirm using the browser's networktab that the header is set

##### Testing the `auto` mode #1

- make sure NO report is published in com_csp
- Go to the Plugin
- Got to the CSP tab in the plugin settings
- Set the `mode` to `auto`
- confirm using the browser's networktab that NO CSP header is set

##### Testing the `auto` mode #2

- Go to com_csp
- publish all collected rules
- Go to the Plugin
- Got to the CSP tab in the plugin settings
- Set the `mode` to `auto`
- confirm using the browser's networktab that the CSP header IS set

### Expected result

All four test cases works

### Actual result

No integration between com_csp & plg_system_httpheaders

### Documentation Changes Required

This is a new feature so docu is needed

### CODE REVIEW

File: plugins/system/httpheaders/httpheaders.php [hunk 1]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
 			$hstsOptions[] = 'preload';
 		}
 
-		$this->app->setHeader('Strict-Transport-Security', implode('; ', $hstsOptions));
+		$this->app->setHeader('Strict-Transport-Security', trim(implode('; ', $hstsOptions)));
+	}
+
+	/**
+	 * Compone the automatic csp header rules based on com_csp / #__csp
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: Compone? Did you mean compose

File: plugins/system/httpheaders/httpheaders.php [hunk 2]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
...
+	 *
+	 * @param  string   $nonce  The System nonce used for script and style tags
+	 *
+	 * @return  array  An array containing the csp rules found in com_csp
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function compileAutomaticCspHeaderRules($nonce): array
+	{
+		// Get the published infos form the database
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: From

File: administrator/components/com_csp/View/Reports/HtmlView.php [hunk 3]
```diff
@@ -150,12 +151,13 @@ protected function addToolbar()
...
-			'a.document_uri' => \JText::_('COM_CSP_HEADING_DOCUMENT_URI'),
-			'a.directive'    => \JText::_('COM_CSP_HEADING_DIRECTIVE'),
-			'a.id'           => \JText::_('JGRID_HEADING_ID')
-		);
+		return [
+			'a.state'        => Text::_('JSTATUS'),
+			'a.blocked_uri'  => Text::_('COM_CSP_HEADING_BLOCKED_URI'),
+			'a.document_uri' => Text::_('COM_CSP_HEADING_DOCUMENT_URI'),
+			'a.directive'    => Text::_('COM_CSP_HEADING_DIRECTIVE'),
+			'a.client   '    => Text::_('JCLIENT'),
```
Comment (Quy) [inline administrator/components/com_csp/View/Reports/HtmlView.php]: Remove spaces

Comment (zero-24) [reply administrator/components/com_csp/View/Reports/HtmlView.php]: fixed with the last commit thanks!

File: administrator/components/com_csp/tmpl/reports/default.php [hunk 4]
```diff
@@ -70,10 +70,13 @@
 								<th>
 									<?php echo JHtml::_('searchtools.sort', 'COM_CSP_HEADING_DIRECTIVE', 'a.directive', $listDirn, $listOrder); ?>
 								</th>
+								<th>
```
Comment (Quy) [inline administrator/components/com_csp/tmpl/reports/default.php]: Change to 8 in `<tfoot>`

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 5]
```diff
@@ -13,6 +13,13 @@ PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
 ; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE="Mode"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_CUSTOM="Custom"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_DETECT="Detect"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_AUTO="Automatic"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Sort in alpha order

File: administrator/components/com_csp/tmpl/reports/default.php (hunk 4 above)
Comment (zero-24) [reply administrator/components/com_csp/tmpl/reports/default.php]: fixed by https://github.com/joomla/joomla-cms/pull/20847/commits/a38817efc63cba3a82b5cdf42e7d7b569229c77d

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 6]
```diff
@@ -13,6 +13,13 @@ PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
 ; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE="Mode"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_CUSTOM="Custom"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_DETECT="Detect"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_AUTO="Automatic"
+; Please only change the URL in the following language string
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_NONCE_ENABLED="<a href='https://en.wikipedia.org/wiki/Cryptographic_nonce' target='_blank'>Nonce</a>"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: we need to have some styling to show this is an external link

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 7]
```diff
@@ -13,6 +13,13 @@ PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
 ; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE="Mode"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_CUSTOM="Custom"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_DETECT="Detect"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_MODE_AUTO="Automatic"
+; Please only change the URL in the following language string
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_NONCE_ENABLED="<a href='https://en.wikipedia.org/wiki/Cryptographic_nonce' target='_blank'>Nonce</a>"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_NONCE_ENABLED_DESC="Enable the whitelist for specific inline scripts using a cryptographic nonce (number used once) for all scripts and styles using the Joomla API. Specifying nonce makes a modern browser ignore 'unsafe-inline' which could still be set for older browsers without nonce support."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: specifying a nonce

File: plugins/system/httpheaders/httpheaders.php [hunk 8]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
...
+			return [];
+		}
+
+		$automaticCspHeader  = [];
+		$cspHeaderCollection = [];
+		$nonceEnabled        = (int) $this->params->get('nonce_enabled', 0);
+
+		foreach ($rows as $row)
+		{
+			// Handle the client information foreach rule
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: for each

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 6 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: hmm the same is true for the other links right and any other place we add links right? I would like to move this than to a dedicated issue when you agree

File: plugins/system/httpheaders/httpheaders.php [hunk 9]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
...
+			if (!$this->app->isClient($row->client))
+			{
+				continue;
+			}
+
+			// Make sure the directive exists as key
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: as a key

File: plugins/system/httpheaders/httpheaders.php [hunk 10]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
...
+			if (!isset($cspHeaderCollection[$row->directive]))
+			{
+				$cspHeaderCollection = array_merge($cspHeaderCollection, array_fill_keys([$row->directive], ''));
+			}
+
+			// Eval or inline lets make sure they still work by adding ' before and after
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: lets us make

File: plugins/system/httpheaders/httpheaders.php [hunk 11]
```diff
@@ -224,6 +312,98 @@ private function setHstsHeader()
...
+			if (in_array($row->blocked_uri, ['unsafe-eval', 'unsafe-inline']))
+			{
+				$row->blocked_uri = "'$row->blocked_uri'";
+			}
+
+			// Whiteliste the blocked_uri for the given directive
```
Comment (brianteeman) [inline plugins/system/httpheaders/httpheaders.php]: Whitelist

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 6 above)
Comment (brianteeman) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: agree. i meant to mention it on the plugin and forgot

Comment (zero-24) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Done: https://github.com/joomla/joomla-cms/issues/20849

**Comment by brianteeman** [comment]
@zero-24 With this component your original idea for providing http headers etc has changed (for the better). However from a UI/UX it is unusual in the joomlasphere in that we have a component that does one part of the job and a plugin that does the rest.

I understand why it is a plugin that does the actual headers but there isnt a good reason (I think) for the creation of the rules to be done in the plugin, especially now we have a component that is detecting some of the rules that you require.

Would it not make more sense to strip out all the configuration from the plugin, other than being enabled/disabled, and putting it all into the component.

Then there would be two views in the component
com_csp?task=report
com_csp?task=setup
and possibly
com_csp?task=options

It will make more sense to view the reports in the component and then apply the recommendations to the headers if we can see the final config in the same component without having to close the component and go to the plugin. It will also be quicker and easier to see what happens all from the same place.

If you look at com_finder as an example then you see that its all done in the component and not the plugins

**Comment by zero-24** [comment]
Hi,

thanks for the feedback. Yes that sounds like a plan. But before I do the development I would like to get @wilsonge check this and tell us a bit more details about the planed design rework and how such a configuration would fit into the new design.

> The design rework is going to give us a separated dashboard and set of components. There is now a more conscious separation between administration components and day to day content components with the split of dashboards. Not all administrators are actually going to be totally aware of this extension.

As I would like to not redo all of this two times. Or worse do something that is completely against that design rework.

Given that this is for now the technical PR with all the features working I would like to get this in even with this know UI/UX improvements and than when we have the details for the design do a separate PR to implement the new design so we have step by step improvements. Would you agree with that?

**Comment by brianteeman** [comment]
Please add rel="noopener noreferrer" to the external links in the language file - see #14881 for the full explanation

**Comment by zero-24** [comment]
> Please add rel="noopener noreferrer" to the external links in the language file - see #14881 for the full explanation

I have done the rest here: https://github.com/joomla/joomla-cms/pull/20866. Thanks!

**Comment by zero-24** [comment]
@wilsonge can we please get your voice on how we should step forward? I would like to see this PR merged so we can start a new PR (who can help me with that?) to include anything in the new design where needed?

**Comment by wilsonge** [comment]
OK I've merged this so we have the integration. We can now see the new design. The aim is that we are going to remove the CSP from the main components menu and leave it only in the system dashboard (along with the various other components that live there)

**Comment by wilsonge** [comment]
I can see Brian's point of view about having a singular place for editing everything - but I'm not totally sure whether it's a good idea or not to be honest. Can I hear thoughts from others?

**Comment by zero-24** [comment]
> OK I've merged this so we have the integration. We can now see the new design. The aim is that we are going to remove the CSP from the main components menu and leave it only in the system dashboard (along with the various other components that live there)

Ok please ping me on Glip on the details so i can do my bit and help implementing it.

**Pull request merged**